### PR TITLE
inital work on cluster source + MOR systematic

### PR DESCRIPTION
**Pull request by elikrause**
This PR documents initial work implementing ClusterSource source and MORMurata systematic. It is far from complete/ready for merging.

### CODE REVIEW

File: firecrown/ccl/sources/sources.py [hunk 1]
```diff
@@ -5,9 +5,12 @@
 import pyccl as ccl
 
 from ..core import Source
+Z_MIN = 0.0
```
Comment (beckermr) [inline firecrown/ccl/sources/sources.py]: Add a new line above this.

File: firecrown/ccl/sources/sources.py [hunk 2]
```diff
@@ -216,3 +219,136 @@ def render(self, cosmo, params, systematics=None):
...
+        `pyccl.NumberCountsTracer`
+    """
+    def __init__(
+            self, bin_data, has_rsd=False,
+            scale=1.0, systematics=None):
+        df = pd.read_csv(bin_data)
+        self.z_min = f['z_min'].values.copy()[0]
+        self.z_max = f['z_max'].values.copy()[0]
+        self.proxy_min = f['proxy_min'].values.copy()[0]
+        self.proxy_max = f['proxy_max'].values.copy()[0]
```
Comment (beckermr) [inline firecrown/ccl/sources/sources.py]: Make these private by adding an underscore in front of them.

File: firecrown/ccl/sources/sources.py [hunk 3]
```diff
@@ -216,3 +219,136 @@ def render(self, cosmo, params, systematics=None):
...
+        self.a_eff = f['area_eff'].values.copy()[0]
```
Comment (beckermr) [inline firecrown/ccl/sources/sources.py]: Ditto here.

File: firecrown/ccl/systematics/cl.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        ----------
+        cosmo : pyccl.Cosmology
+            A pyccl.Cosmology object.
+        params : dict
+            A dictionary mapping parameter names to their current values.
+        source : a source object
+            The source to which apply the MOR model.
+        """
+        #How do I pass this MOR systematic back to source?
+        source.bias_ = source.int_p_dz_dm_dproxy(cosmo, params, params, self, weight = ccl.halo_bias)
```
Comment (beckermr) [inline firecrown/ccl/systematics/cl.py]: Use *= here.

File: firecrown/ccl/sources/sources.py [hunk 5]
```diff
@@ -216,3 +219,136 @@ def render(self, cosmo, params, systematics=None):
...
+        _a = 1./(1+self.z_)
+        self.dndz_ = ccl.h_over_h0(cosmo,_a)*(ccl.comoving_radial_distance(cosmo, _a)**2)
+
+        self.dndz_interp = Akima1DInterpolator(
+            self.z_ , self.dndz_)
+        self.scale_ = self.scale
+
+        for systematic in self.systematics:
+            systematics[systematic].apply(cosmo, params, self)
+        #need to check that self.systematics includes some type of MOR
```
Comment (beckermr) [inline firecrown/ccl/sources/sources.py]: Make this a TODO. We can add checks later.

File: firecrown/ccl/sources/sources.py [hunk 6]
```diff
@@ -216,3 +219,136 @@ def render(self, cosmo, params, systematics=None):
...
+
+        #what's happening here?
+        self.bias_ = np.ones_like(self.z_) * params[self.bias]
```
Comment (beckermr) [inline firecrown/ccl/sources/sources.py]: This line should go before the loop over systematics.

File: firecrown/ccl/sources/sources.py [hunk 7]
```diff
@@ -216,3 +219,136 @@ def render(self, cosmo, params, systematics=None):
...
+        df = pd.read_csv(bin_data)
+        self.z_min = f['z_min'].values.copy()[0]
+        self.z_max = f['z_max'].values.copy()[0]
+        self.proxy_min = f['proxy_min'].values.copy()[0]
+        self.proxy_max = f['proxy_max'].values.copy()[0]
+        self.a_eff = f['area_eff'].values.copy()[0]
+        self.has_rsd = has_rsd
+        self.systematics = systematics or []
+        self.scale = scale
+        self.bias_ = 0
```
Comment (beckermr) [inline firecrown/ccl/sources/sources.py]: Remove this line.

File: firecrown/ccl/sources/sources.py (hunk 6 above)
Comment (beckermr) [reply firecrown/ccl/sources/sources.py]: Remove params[self.bias]

File: firecrown/ccl/sources/sources.py [hunk 8]
```diff
@@ -216,3 +219,136 @@ def render(self, cosmo, params, systematics=None):
...
+    Methods
+    -------
+    render : apply systematics to this source and build the
+        `pyccl.NumberCountsTracer`
+    """
+    def __init__(
+            self, bin_data, has_rsd=False,
+            scale=1.0, systematics=None):
+        df = pd.read_csv(bin_data)
+        self.z_min = f['z_min'].values.copy()[0]
```
Comment (beckermr) [inline firecrown/ccl/sources/sources.py]: Let's add `self.bin_data = bin_data`.

File: firecrown/ccl/sources/sources.py [hunk 9]
```diff
@@ -216,3 +219,136 @@ def render(self, cosmo, params, systematics=None):
...
+        self.z_max = f['z_max'].values.copy()[0]
+        self.proxy_min = f['proxy_min'].values.copy()[0]
+        self.proxy_max = f['proxy_max'].values.copy()[0]
+        self.a_eff = f['area_eff'].values.copy()[0]
+        self.has_rsd = has_rsd
+        self.systematics = systematics or []
+        self.scale = scale
+        self.bias_ = 0
+    def int_p_dz_dm_dproxy(self, cosmo, params, mor, weight=None):
```
Comment (beckermr) [inline firecrown/ccl/sources/sources.py]: Add a new line above this.

File: firecrown/ccl/sources/sources.py [hunk 10]
```diff
@@ -216,3 +219,136 @@ def render(self, cosmo, params, systematics=None):
...
+        """
+        evaluate \int dz n(z) \int dM n(M,z) weight(M,z) \int dproxy P(proxy|M,z)
```
Comment (beckermr) [inline firecrown/ccl/sources/sources.py]: Docs start like this 

```python
def func():
    """This is my func.
    """
```

File: firecrown/ccl/sources/sources.py [hunk 11]
```diff
@@ -216,3 +219,136 @@ def render(self, cosmo, params, systematics=None):
...
+        if weight is None:
+            weight = lambda cosmo, M, a : 1.0
+
+        if weight is not None:
+            norm = self.int_dz_dM_dproxy(cosmo, params, mor)
+        else:
+            norm = 1
+
+        def _integrand_p_dz_dm_dproxy(z, lgm):
+            math.exp(lgm)*self.dndz(z)*mor.int_p_dproxy(proxy, m,z)*ccl.massfunc(cosmo, math.exp(lgm),1/(1+z))*weight(cosmo, math.exp(lgm),1/(1+z))
```
Comment (beckermr) [inline firecrown/ccl/sources/sources.py]: Use `np.exp` not `math.exp`.

File: firecrown/ccl/sources/sources.py [hunk 12]
```diff
@@ -216,3 +219,136 @@ def render(self, cosmo, params, systematics=None):
...
+        params : dict
+            A dictionary mapping parameter names to their current values.
+        mor: systematic
+            A Mass-Observable relation systematic. 
+        weight : function(cosmo, m, a)
+            An optional weight function with arguments cosmo, halo mass, scale factor.
+        """
+
+        if weight is None:
+            weight = lambda cosmo, M, a : 1.0
```
Comment (beckermr) [inline firecrown/ccl/sources/sources.py]: `M` -> `m`

File: firecrown/ccl/sources/sources.py (hunk 11 above)
Comment (beckermr) [reply firecrown/ccl/sources/sources.py]: Missed the return statement.

Comment (beckermr) [reply firecrown/ccl/sources/sources.py]: Use the interpolated redshift dist.

File: firecrown/ccl/systematics/cl.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        Parameters
+        ----------
+        cosmo : pyccl.Cosmology
+            A pyccl.Cosmology object.
+        params : dict
+            A dictionary mapping parameter names to their current values.
+        source : a source object
+            The source to which apply the MOR model.
+        """
+        #How do I pass this MOR systematic back to source?
```
Comment (beckermr) [inline firecrown/ccl/systematics/cl.py]: The mor can be passed as `self`.

**Comment by beckermr** [comment]
I am thinking about this more. I think maybe we want to store the MOR as a property that gets set on the source by the systematic. For example, the fiducial one could be a perfect mass tracer and then systematics can replace the MOR.

File: firecrown/ccl/sources/tests/test_cl_source.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+f = open('dummy_cl_bin_data.csv', 'w')
+print(','.join(bin_data.keys())+'\n'+
+        ','.join([str(i) for i in bin_data.values()]),
+        file=f)
+f.close()
+cl = ClusterSource('dummy_cl_bin_data.csv')
+os.system('rm dummy_cl_bin_data.csv')
+print(cl)
```
Comment (beckermr) [inline firecrown/ccl/sources/tests/test_cl_source.py]: This needs to be done in a temporary directory made within the test. See `tempfile.TemporaryDirectory()` and please use a context manager.

File: firecrown/ccl/sources/tests/test_cl_source.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
+from ..sources import ClusterSource
+import os
+bin_data = {
```
Comment (beckermr) [inline firecrown/ccl/sources/tests/test_cl_source.py]: You want this data located in the test.

File: firecrown/ccl/sources/tests/test_cl_source.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    'z_min':0.1,
+    'z_max':1.0,
+    'proxy_min':20,
+    'proxy_max':40,
+    'area_eff':10,
+}
+
+f = open('dummy_cl_bin_data.csv', 'w')
```
Comment (beckermr) [inline firecrown/ccl/sources/tests/test_cl_source.py]: Always use a context manager when opening files.

File: firecrown/ccl/sources/tests/test_cl_source.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+print(','.join(bin_data.keys())+'\n'+
+        ','.join([str(i) for i in bin_data.values()]),
+        file=f)
+f.close()
+cl = ClusterSource('dummy_cl_bin_data.csv')
+os.system('rm dummy_cl_bin_data.csv')
```
Comment (beckermr) [inline firecrown/ccl/sources/tests/test_cl_source.py]: This call is not safe. Use a tempdir as stated below.

**Comment by vitenti** [comment]
We are reorganizing the Cluster counts likelihood in Firecrown. I'm closing this PR and any useful code will be copied to the new structure.

**Pull request closed without merging**